tools/hash_reqs.py

```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env python
+
+from __future__ import print_function
+
+import os
+import sys
+import argparse
+import hashlib
+
+def expand_reqs_helper(fpath, visited):
+    if fpath in visited:
+        raise Exception("Cyclic dependency between requirements files")
+
+    curr_dir = os.path.dirname(fpath)
+    result = [] # type: List[str]
+
+    for line in open(fpath):
+        dep = line.split(" #", 1)[0].strip() # remove comments and strip whitespace
+        if dep:
+            if dep.startswith('-r'):
+                child = os.path.join(curr_dir, dep[3:])
+                result += expand_reqs_helper(child, visited.union({fpath}))
+            else:
+                result.append(dep)
+    return result
+
+def expand_reqs(fpath):
+    """
+    Returns a sorted list of unique dependencies specified by the requirements file `fpath`.
+    Removes comments from the output and recursively visits files specified inside `fpath`.
+    `fpath` can be either an absolute path or a relative path.
+    """
+    absfpath = os.path.abspath(fpath)
+    output = expand_reqs_helper(absfpath, set())
```

Comment:
Oh you're right. My algo will fail in a diamond relationship. I'm not detecting cycles correctly.

Pip doesn't allow cycles (it gives a `maximum recursion depth exceeded` error) so I thought it would be better to throw an error on detecting a cycle.

Detecting a cycle will not really benefit us and it'll complicate code. So I'll abandon that idea. I'll just do nothing when I come across a visited file.